Fix Part of #13162: Add argument schema for ExplorationCompleteEventHandler, ExplorationMaybeLeaveHandler, SolutionHitEventHandler classes

SD-13 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of  #13162.
2. This PR does the following: This PR adds schema for ExplorationCompleteEventHandler, ExplorationMaybeLeaveHandler, SolutionHitEventHandler classes.

## Essential Checklist

- [X] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [X] The linter/Karma presubmit checks have passed locally on your machine.
- [X] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [X] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

![Screenshot from 2021-11-09 13-22-55](https://user-images.githubusercontent.com/89520981/140941918-f4ae1258-f9d7-469b-8554-f9305775000c.png)

![Screenshot from 2021-11-09 12-18-25](https://user-images.githubusercontent.com/89520981/140942043-796e29da-107d-4324-a3ca-39d029b7d1c7.png)
![Screenshot from 2021-11-09 12-17-45](https://user-images.githubusercontent.com/89520981/140942205-ea218301-3ec2-42c4-86bc-8129df47c5c4.png)
![Screenshot from 2021-11-09 12-17-54](https://user-images.githubusercontent.com/89520981/140942237-3a480c66-aa4a-4fc6-8034-b3501f642efc.png)



<!--
Add videos/screenshots of the user-facing interface in various display sizes (mainly phone, tablet, and desktop display size) to demonstrate that the changes made in this PR work correctly.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

SD-13 committed:
adds schema for few handlers

SD-13 committed:
updated according to lint checks

SD-13 commented:
@aks681 @vojtechjelinek Please review my PR and tell me what changes are needed. Thanks

SD-13 commented on core/controllers/reader.py:
```diff
@@ -608,18 +610,44 @@ def post(self, exploration_id):
 class SolutionHitEventHandler(base.BaseHandler):
     """Tracks a learner clicking on the 'View Solution' button."""
 
+    URL_PATH_ARGS_SCHEMAS = {
+        'exploration_id': {
+            'schema': editor.SCHEMA_FOR_EXPLORATION_ID
+        }
```
Is it ok or should I change it?

vojtechjelinek replied:
Ok

vojtechjelinek commented on core/controllers/reader.py:
```diff
@@ -608,18 +610,44 @@ def post(self, exploration_id):
+            },
+            'state_name': {
+                'schema': {
+                    'type': schema_utils.SCHEMA_TYPE_BASESTRING
```
I think we usually do not use constants for these simple strings (we do not use them in schemas, in other parts of the codebase we use them).

SD-13 replied:
done.

SD-13 replied:
I have replaced those constants. Please tell me If it needs any other changes.

vojtechjelinek commented on core/controllers/reader.py:
```diff
@@ -608,18 +610,44 @@ def post(self, exploration_id):
-            exploration_id, self.payload.get('exploration_version'),
-            self.payload.get('state_name'), self.payload.get('session_id'),
-            self.payload.get('time_spent_in_state_secs'))
+            exploration_id, self.normalized_payload.get('exploration_version'),
+            self.normalized_payload.get('state_name'),
+            self.normalized_payload.get('session_id'),
+            self.normalized_payload.get('time_spent_in_state_secs'))
```
```suggestion
            exploration_id, 
            self.normalized_payload.get('exploration_version'),
            self.normalized_payload.get('state_name'),
            self.normalized_payload.get('session_id'),
            self.normalized_payload.get('time_spent_in_state_secs'))
```

SD-13 replied:
done

SD-13 replied:
Please let me know if I missed any changes here.

vojtechjelinek commented on core/controllers/reader.py:
```diff
@@ -629,6 +657,47 @@ class ExplorationCompleteEventHandler(base.BaseHandler):
+                    'type': schema_utils.SCHEMA_TYPE_FLOAT
+                }
+            },
+            'params': {
+                'schema': {
+                    'type': 'object_dict',
+                    'validation_method': (
+                        domain_objects_validator.validate_params),
+                }
+            }
```
Why are the params send like this?

vojtechjelinek replied:
Can some more structured manner be used?

SD-13 replied:
I think here I picked the wrong type. Later I found `params` arg in the domain layer and I have updated the schema accordingly. Please have a look and let me know if any changes are needed.

vojtechjelinek reviewed (changes requested):
Thanks! Left a few comments.

SD-13 committed:
replaced constants and added few suggested changes

SD-13 committed:
removed an unused import

SD-13 committed:
updated schema for params arg

SD-13 committed:
updated according to lint checks

SD-13 commented:
@vojtechjelinek PTAL. Thanks : )

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into Adds_schema_for_handlers_4

SD-13 committed:
updated few schema and tests

SD-13 committed:
updated with lint checks

vojtechjelinek commented on core/controllers/reader.py:
```diff
@@ -608,18 +609,45 @@ def post(self, exploration_id):
+            'time_spent_in_state_secs': {
+                'schema': {
+                    'type': 'float'
+                }
```
Maybe this needs to be positive value?

SD-13 replied:
yes, I have added rules here.

vojtechjelinek commented on core/controllers/reader.py:
```diff
@@ -608,18 +609,45 @@ def post(self, exploration_id):
+                'schema': {
+                    'type': 'basestring'
+                }
+            },
```
Are there any length rules for state name?

SD-13 replied:
yes. I have added. Please have a look.

vojtechjelinek commented on core/controllers/reader.py:
```diff
@@ -629,6 +657,50 @@ class ExplorationCompleteEventHandler(base.BaseHandler):
+                'schema': {
+                    'type': 'basestring'
+                }
+            },
```
Again, any length limit?

SD-13 replied:
done.

vojtechjelinek commented on core/controllers/reader.py:
```diff
@@ -629,6 +657,50 @@ class ExplorationCompleteEventHandler(base.BaseHandler):
+                'schema': {
+                    'type': 'float'
+                }
+            },
```
Again, should be positive.

SD-13 replied:
done.

vojtechjelinek commented on core/controllers/reader.py:
```diff
@@ -683,6 +752,46 @@ class ExplorationMaybeLeaveHandler(base.BaseHandler):
+                    'type': 'basestring'
+                },
+                'default_value': None
+            },
```
More rules for `collection_id` should be possible.

SD-13 replied:
done.

vojtechjelinek commented on core/controllers/reader.py:
```diff
@@ -683,6 +752,46 @@ class ExplorationMaybeLeaveHandler(base.BaseHandler):
+                    'type': 'basestring'
+                }
+            },
+            'client_time_spent_in_secs': {
+                'schema': {
+                    'type': 'float'
+                }
+            },
```
Should be possitive.

SD-13 replied:
done.

vojtechjelinek commented on core/controllers/reader.py:
```diff
@@ -683,6 +752,46 @@ class ExplorationMaybeLeaveHandler(base.BaseHandler):
+            'version': {
+                'schema': editor.SCHEMA_FOR_VERSION
+            },
+            'state_name': {
+                'schema': {
+                    'type': 'basestring'
+                }
+            },
```
Some length limit?

SD-13 replied:
done

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into Adds_schema_for_handlers_4

SD-13 committed:
added rules in few handler args

SD-13 commented:
I have updated the PR. Please have a look and let me know if any changes are needed. @vojtechjelinek PTAL.

vojtechjelinek commented on core/controllers/reader.py:
```diff
@@ -629,6 +665,58 @@ class ExplorationCompleteEventHandler(base.BaseHandler):
+                    }]
+                }
+            },
+            'params': {
+                'schema': {
+                    'type': 'dict',
+                    'properties': []
+                }
```
Please add an explanatory comment, why are params set like this.

SD-13 replied:
I have navigated both the frontend and backend parts and I found that they are customizable for any exploration (they cannot be used in newer explorations). So, Probably the type should be `dict` here.

SD-13 replied:
1. Later I found `params` handler arg in the domain layer, which means we should choose the type `object_dict`. 
2. Then by your suggestion I got to know what params is. I found in the backend that it follows `dict` type and I took `validation_method` to add rules for it.

vojtechjelinek commented on core/controllers/reader.py:
```diff
@@ -683,6 +768,58 @@ class ExplorationMaybeLeaveHandler(base.BaseHandler):
+                    }]
+                }
+            },
+            'params': {
+                'schema': {
+                    'type': 'dict',
+                    'properties': []
+                }
+            }
```
ditto as above

SD-13 replied:
same as above

SD-13 replied:
Please take a look [here](https://github.com/oppia/oppia/pull/14208#discussion_r757932663).

vojtechjelinek reviewed (changes requested):
Thanks! Left a few comments.

SD-13 committed:
adds rules in params schema

SD-13 committed:
resolve merge conflict

SD-13 committed:
updated the schema

SD-13 committed:
removed handler names from HANDLER_CLASS_NAMES_WHICH_STILL_NEED_SCHEMAS list

SD-13 committed:
updated a test for params

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into Adds_schema_for_handlers_4

SD-13 committed:
adds a test for params

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into Adds_schema_for_handlers_4

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into Adds_schema_for_handlers_4

srijanreddy98 committed:
Fixes part of #9749: Migrate few instances of angular-html-bind (#14263)

* Revert "Revert "Fix part of #9749: Create a component that shows rte-output-display and use it instead of angular-html-bind. (#13229)" (#13361)"

This reverts commit 792de4596996527a696d0041e0eff2b8384eda30.

* Fix issues

* Fix spelling mistakes in comment

* Merge with upstream/develop

* Address review comments:
  1. Move oppia-rte-output.spec from the list of ignored files for innerHTML check in eslintrc and move to inline disables.
  2. Change the CODEOWNER listing from oppia-rte-parser.service.ts to oppia-rte-parser.service*.ts to include the spec file.
  3. Add refernce to the note at the top of file for all value.replace ops.
  4. Miscellaneous changes.

Co-authored-by: Vasa Srijan Reddy <redvasa@amazon.com>

SD-13 committed:
adds test for params value

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into Adds_schema_for_handlers_4

SD-13 committed:
adds test for params invalid data

SD-13 committed:
fixed lint checks

SD-13 committed:
updated rules

SD-13 commented:
@Nik-09 PTAL.

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into Adds_schema_for_handlers_4

Nik-09 commented on core/controllers/domain_objects_validator.py:
```diff
@@ -178,3 +178,13 @@ def validate_aggregated_stats(aggregated_stats):
+    Args:
+        params: dict. Data that needs to be validated.
+    """
+    for value in params.values():
```
1. @SD-13 Here you should verify the type of key to be string instead of value -- based on [this comment](https://github.com/oppia/oppia/blob/8b1986ef0746ebec1d52c3243c078cc58bdd15a0/core/templates/pages/exploration-player-page/services/stats-reporting.service.ts#L352) it is a statename.

2. Also just for confirmation, if keys are state name then its values can be a State Object in dict format. If this is the case can we use validate method for state object?

SD-13 replied:
Please take a look [here](https://github.com/oppia/oppia/pull/14208/files#r757851606).

Nik-09 reviewed (commented):
@SD-13 Left comments for confirmation.

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into Adds_schema_for_handlers_4

SD-13 committed:
adds check for invalid key

SD-13 commented on core/controllers/domain_objects_validator.py:
```diff
@@ -178,3 +178,14 @@ def validate_aggregated_stats(aggregated_stats):
+    Args:
+        params: dict. Data that needs to be validated.
+    """
+    for key in params.keys():
+        if not isinstance(key, str):
+            raise Exception('Excepted str, received %s' % key)
```
1. @Nik-09 here I have tried to add a check for `key` and I have also added a test in `core/controllers/reader_test.py` with non-string type data. In the test, I was expecting status code 400 but I am getting 200. what you would suggest to fix the test?
2.  [Here](https://github.com/oppia/oppia/blob/develop/core/storage/statistics/gae_models.py#L453) `params` is following `dict[str, str]` so, can we keep the rules for values?

Nik-09 replied:
For point 1. -- You have to debug by adding print logs in every stage of the architecture, this is the general way you can solve a bug. Also I think you need to add backend tests in domain_objects_validator_test.py file.

For point 2 -- This is amibigious as in frontend params type is mentioned as Object and here in the backend it is mentioned as str (also with a note of confusion as given [here](https://github.com/oppia/oppia/blob/develop/core/storage/statistics/gae_models.py#L441))
As far I remember you reproduced e2e tests failures because  of type str but not sure.
@vojtechjelinek Can you suggest something for this ?

SD-13 replied:
@Nik-09 I have added a backend test in the `domain_objects_validator_test.py` file. Please have a look.

SD-13 commented:
@Nik-09 I have updated the PR. Please have a look. Thanks

Nik-09 commented on core/controllers/domain_objects_validator.py:
```diff
@@ -186,5 +186,6 @@ def validate_params_dict(params):
     """
-    for value in params.values():
-        assert isinstance(value, str), ('Expected str, got %s' % value)
+    for key in params.keys():
```
Confirm that the type of params to be dict.

SD-13 replied:
yes, I found it as `dict` in many files.

SD-13 committed:
adds a test

SD-13 commented on core/controllers/domain_objects_validator.py:
```diff
@@ -178,3 +178,14 @@ def validate_aggregated_stats(aggregated_stats):
+        params: dict. Data that needs to be validated.
+    """
+    for key in params.keys():
+        if not isinstance(key, str):
+            raise Exception('Excepted str, received %s' % key)
```
@Nik-09 After adding print logs I found that though I am passing non-string type data this condition still remains false.

SD-13 committed:
removed tests and updated the rules

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into Adds_schema_for_handlers_4

SD-13 committed:
added test for the rules

SD-13 commented:
@vojtechjelinek PTAL. Thanks

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into Adds_schema_for_handlers_4

Nik-09 commented on core/controllers/domain_objects_validator.py:
```diff
@@ -178,3 +178,13 @@ def validate_aggregated_stats(aggregated_stats):
+        params: dict. Data that needs to be validated.
+    """
+    if not isinstance(params, dict):
+        raise Exception('Excepted dict, received %s' % params)
```
I was asking to add a validation inorder to expect the type of the params to be dict, as you have done here.
Also verify that the keys in params dict are of string type.

Also can you modify the type annotation [here](https://github.com/oppia/oppia/blob/develop/core/storage/statistics/gae_models.py#L453)

SD-13 replied:
Yes, I found in many places that it is taking string-type keys.
previously I tried to add rules for both keys and values individually and that was causing lots of e2e failures and it seemed like they were transformed into string types before passing into the validator.
Also, I did not find anything that violates [them](https://github.com/oppia/oppia/blob/4e3d204fad3be36873fea307ae9c56ae1cf51a26/core/storage/statistics/gae_models.py#L453). then what did you ask me to modify? Can you please explain

Nik-09 replied:
I think you can ignore this and leave it for now

SD-13 replied:
Okay.

Nik-09 reviewed (commented):
Thank you @SD-13, added one comment.

vojtechjelinek reviewed (approved):
LGTM!

aks681 reviewed (approved):
Lgtm as codeowner

SD-13 commented:
Thanks,  @aks681 @vojtechjelinek : )

Nik-09 reviewed (approved):
@SD-13  Update the PR with upstream.

SD-13 committed:
Merge remote-tracking branch 'upstream/develop' into Adds_schema_for_handlers_4

SD-13 commented:
> @SD-13 Update the PR with upstream.

Done.
